### PR TITLE
feat(gui): async loading of thumbnails

### LINKED ISSUE #257: make timeline thumbnail cache lazy
## User story
1. As a user
2. I want that the timeline works without lags all the time
3. So that the user can navigate through the video

## Acceptance criteria
* the timeline thumbnail cache should be loaded lazy
* as long as there is no thumbnail loaded there should be a grey placeholder

## Definition of done (DoD)
* Create feature branch
* Implement the functionality
* Opened pull request and assigned reviewer
* If necessary: required changes are implemented
* Delete feature branch
* If necessary: Component Test has been written
* Code review has been completed and code has been merged

## DoD general criteria
* Feature has been fully implemented
* Feature has been merged into the mainline
* All acceptance criteria were met
* Product owner approved features
* All tests are passing
* Developers agreed to release

### PR DESCRIPTION
**Pull request by akriese**
Closes #257 

This PR changes the loading of the thumbnails.
There also some refactorings being made.
The FrameNavigation is split into a Navigation and a FramGrabber part.

### CODE REVIEW

**Comment by akriese** [comment]
rebased

**Comment by simonsasse** [comment]
works like a charm :)

File: GUI/src/main/kotlin/logic/FrameGrabber.kt [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    fun getCurrentVideoFrame(index: Int): ImageBitmap {
+        if (videoCurrentFrames[index] != -1) {
+            videoCurrentGrabber.setVideoFrameNumber(videoCurrentFrames[index])
+            return getBitmap(videoCurrentGrabber)
+        } else {
+            return deletionBitmap
+        }
+    }
+
+    fun getDiffVideoFrame(index: Int): ImageBitmap {
```
Comment (simonsasse) [inline GUI/src/main/kotlin/logic/FrameGrabber.kt:77]: doc

File: GUI/src/main/kotlin/logic/caches/ThumbnailCache.kt [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package logic.caches
 
 import androidx.compose.ui.graphics.ImageBitmap
+import org.bytedeco.opencv.opencv_core.Mutex
```
Comment (simonsasse) [inline GUI/src/main/kotlin/logic/caches/ThumbnailCache.kt]: shouldn't this be `import kotlinx.coroutines.sync.Mutex` ? Anton said something about this.

File: GUI/src/main/kotlin/logic/caches/ThumbnailCache.kt [hunk 3]
```diff
@@ -23,9 +25,12 @@ class ThumbnailCache(val maxCacheSize: Int, val getImages: (Int) -> List<ImageBi
      */
     fun get(index: Int): List<ImageBitmap> {
         recentlyUsed.remove(index)
+        println(index)
```
Comment (simonsasse) [inline GUI/src/main/kotlin/logic/caches/ThumbnailCache.kt]: If using the Kotlin mutex, this line can be removed, said Anton :)

File: GUI/src/main/kotlin/ui/components/diffScreen/timeline/ThumbnailPreview.kt [hunk 4]
```diff
@@ -36,6 +38,7 @@ import ui.components.general.AutoSizeText
  */
 @Composable
```
Comment (simonsasse) [inline GUI/src/main/kotlin/ui/components/diffScreen/timeline/ThumbnailPreview.kt:39]: update doc params: add frameGrabber

File: GUI/src/main/kotlin/ui/components/diffScreen/timeline/ThumbnailPreview.kt [hunk 5]
```diff
@@ -118,7 +122,8 @@ fun AsyncDiffColumn(
     val scope = rememberCoroutineScope()
 
     LaunchedEffect(thumbnailCache, index) {
-        scope.launch {
+        scope.launch(Dispatchers.IO) {
+            println("thumbnail $index")
```
Comment (simonsasse) [inline GUI/src/main/kotlin/ui/components/diffScreen/timeline/ThumbnailPreview.kt:126]: remove println

File: GUI/src/main/kotlin/ui/components/diffScreen/timeline/ThumbnailPreview.kt [hunk 6]
```diff
@@ -163,6 +168,7 @@ fun AsyncDiffColumn(
 @Composable
```
Comment (simonsasse) [inline GUI/src/main/kotlin/ui/components/diffScreen/timeline/ThumbnailPreview.kt:168]: update doc params: add frameGrabber

File: GUI/src/main/kotlin/ui/components/diffScreen/timeline/Timeline.kt [hunk 7]
```diff
@@ -24,7 +25,10 @@ import ui.components.general.TitleWithInfo
  * @return [Unit]
```
Comment (simonsasse) [inline GUI/src/main/kotlin/ui/components/diffScreen/timeline/Timeline.kt:26]: update doc params: add frameGrabber

File: GUI/src/main/kotlin/ui/components/diffScreen/timeline/Timeline.kt [hunk 8]
```diff
@@ -34,6 +38,8 @@ fun Timeline(navigator: FrameNavigation) {
     // set the modifier applied to all timeline components
     val generalModifier = Modifier.fillMaxWidth(fillWidth)
 
+    println("timeline")
```
Comment (simonsasse) [inline GUI/src/main/kotlin/ui/components/diffScreen/timeline/Timeline.kt:44]: remove printLn

File: GUI/src/main/kotlin/ui/screens/DiffScreen.kt [hunk 9]
```diff
@@ -46,9 +47,12 @@ fun DiffScreen(state: MutableState<AppState>) {
     val scope = rememberCoroutineScope()
     val navigator = FrameNavigation(state, scope)
     val showConfirmationDialog = remember { mutableStateOf(false) }
+    val frameGrabber = FrameGrabber(state)
+
     DisposableEffect(Unit) {
         onDispose {
             navigator.close()
+            frameGrabber.close()
```
Comment (simonsasse) [inline GUI/src/main/kotlin/ui/screens/DiffScreen.kt:55]: The line above closes the frameGrabber already. Or not, its closing another instance that lives within the navigator instance, right?

File: GUI/src/main/kotlin/logic/DiffSequenceInfo.kt [hunk 10]
```diff
@@ -0,0 +1,53 @@
+package logic
+
+import algorithms.AlignmentElement
+
+class DiffSequenceInfo(private val diffSequence: Array<AlignmentElement>) {
```
Comment (simonsasse) [inline GUI/src/main/kotlin/logic/DiffSequenceInfo.kt:5]: document class

File: GUI/src/main/kotlin/ui/components/diffScreen/timeline/StatisticalInformation.kt [hunk 11]
```diff
@@ -20,10 +20,7 @@ import util.ColorEncoding
  * @param navigator [FrameNavigation] containing the navigation logic.
  */
 @Composable
```
Comment (simonsasse) [inline GUI/src/main/kotlin/ui/components/diffScreen/timeline/StatisticalInformation.kt:22]: update docu params to diffSequenceInfo

File: GUI/src/main/kotlin/logic/ImageSaveUtil.kt [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    // create the collage
+    val collage =
+        BufferedImage(
+            width * 3 + border * 2,
+            frameGrabber.height + titleHeight,
+            BufferedImage.TYPE_INT_RGB,
+        )
+    val g = collage.createGraphics()
+    // fill the background with white
+    g.color = Color.WHITE
```
Comment (simonsasse) [inline GUI/src/main/kotlin/logic/ImageSaveUtil.kt:41]: could use colorscheme of our software (Material3)

File: GUI/src/main/kotlin/logic/ImageSaveUtil.kt [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+            frameGrabber.getCurrentVideoFrame(diffIndex),
+        )
+    for (item in images) {
+        val img = item.toAwtImage()
+        g.drawImage(img, xOffset, titleHeight, null)
+        xOffset += width + border
+    }
+
+    // draw the titles
+    g.color = Color.BLACK
```
Comment (simonsasse) [inline GUI/src/main/kotlin/logic/ImageSaveUtil.kt:58]: could use colorscheme of our software (Material3)

**Review by simonsasse — DISMISSED** [review]

File: GUI/src/main/kotlin/logic/ImageSaveUtil.kt (hunk 12 above)
Comment (akriese) [reply GUI/src/main/kotlin/logic/ImageSaveUtil.kt:41]: lets do this in another PR :)

File: GUI/src/main/kotlin/logic/ImageSaveUtil.kt (hunk 13 above)
Comment (akriese) [reply GUI/src/main/kotlin/logic/ImageSaveUtil.kt:58]: lets do this in another PR :)

File: GUI/src/main/kotlin/ui/screens/DiffScreen.kt (hunk 9 above)
Comment (akriese) [reply GUI/src/main/kotlin/ui/screens/DiffScreen.kt:55]: I just removed the frameGrabber created in navigator as it serves no purpose anymore

**Comment by akriese** [comment]
@simonsasse I corrected the offset for local density. Can you try it out?

**Comment by zino212** [comment]
@akriese works on my machine :)

**Review by simonsasse — APPROVED** [review]
Sweet

**Pull request merged**